pink/barriers/self_collision_barrier.py

```diff
@@ -0,0 +1,196 @@
+    :math:`d(p^1_i, p^2_i)` is the distance between collision bodies in the pair,
+    and :math:`d_{min}` is minimal distance between any collision bodies.
+
+    Note:
+        The amount of evaluated collision pairs might not be equal to number of all collision pairs.
+        If the amount is less, then only the closest collision pairs will be used.
+
+    Attributes:
+        d_min: Minimum distance between collision pairs.
+    """
+
+    d_min: float
+
+    def __init__(
+        self,
+        n_collision_pairs: int,
+        gain: Union[float, np.ndarray] = 1.0,
+        safe_displacement_gain: float = 1.0,
+        d_min: float = 0.02,
+    ):
+        """Initialization of the SelfCollisionBarrier.
+
+        Args:
+            n_collision_pairs: Number of collision pairs.
+                Note that the number of collision pairs don't have to be equal to
+                the total number of collision pairs in the model. If it is less,
+                than only the closest collision pairs will be used.
+            gain: Barrier gain. Defaults to 1.0.
+            safe_displacement_gain: gain for the safe backup displacement.
+                cost term. Defaults to 1.0.
+            d_min: Minimum distance between any collision pairs.
+                Defaults to 0.02.
+        """
+        super().__init__(
+            dim=n_collision_pairs,
+            gain=gain,
+            safe_displacement_gain=safe_displacement_gain,
+        )
+
+        self.d_min = d_min
+        self.__q_prev = None
+
+    def compute_barrier(self, configuration: Configuration) -> np.ndarray:
+        r"""Compute the value of the barrier function.
+
+        The barrier function is computed as distnace between all collision pairs.
+
+        The barrier function is defined as:
+
+        .. math::
+
+            h(q) = \begin{bmatrix} \ldots \\ d(p^1_i, p^2_i) - d_{min}^2 \\ \ldots \end{bmatrix}
+                \quad \forall i \in 0 \ldots N
+
+        where :math:`N` is number of collision pairs, :math:`p^k_i` is the :math:`k-th` body in :math:`i`-th collision pair,
+        :math:`d(p^1_i, p^2_i)` is the distance between collision bodies in the pair,
+        and :math:`d_{min}` is minimal distance between any collision bodies.
+
+        Args:
+            configuration: Robot configuration :math:`q`.
+
+        Returns:
+            Value of the barrier function
+                 :math:`h(q)`.
+        """
+        return np.array(
+            [
+                configuration.collision_data.distanceResults[k].min_distance
+                - self.d_min
+                for k in range(
+                    len(configuration.collision_model.collisionPairs)
+                )
+            ]
+        )
+
+    def compute_jacobian(self, configuration: Configuration) -> np.ndarray:
+        r"""Compute the Jacobian matrix of the barrier function.
+
+        The jacobian matrix could be represented as stacked gradients of each
```

Comment:
nit (there are a few other occurrences of this below):

```suggestion
        The Jacobian matrix could be represented as stacked gradients of each
```